framework/web/Application.php

```diff
@@ -118,6 +188,64 @@ public function handleRequest($request)
 
     private $_homeUrl;
 
+    /**
+     * @param string $route
+     * @return array|bool
+     */
+    public function createController($route)
+    {
+        $parts = parent::createController($route);
+        if (is_array($parts)) {
+            [$controller, $action] = $parts;
+            if (method_exists($controller, 'middleware')) {
+                foreach ($controller->middleware() as $middleware) {
+                    $this->addMiddleware($middleware);
+                }
+            }
+            $this->callMiddlewareStack($action);
+        }
+        return $parts;
+    }
+
+    /**
+     * Sort middleware by priority and execute
+     *
+     * @param $action string|null
+     * @throws MiddlewareException
+     */
+    protected function callMiddlewareStack($action)
+    {
+        usort($this->middlewareStack, function ($a, $b) {
+            if ($a['priority'] == $b['priority']) {
+                return 0;
+            }
+            return $a['priority'] > $b['priority'] ? -1 : 1;
```

Comment:
I know it's only draft, but perhaps spaceship operator will be nice here due to php7.1+ requirements?